File: src/lib/time-mapping/game-time-map.ts
/**
 * Global game-time mapping (ADR 0002, PRD s3).
 *
 * A hockey game is recorded as N ordered chapter files (a GoPro splits a long
 * recording at ~4 GB). The whole system - player, cut-worker, whistle detector -
 * reasons in a single coordinate: the *global game-time offset*, in seconds since
 * the start of the game, independent of which chapter a moment falls in.
 *
 * This module is the one central, pure, DB-free conversion between that global
 * offset and a `(source file, local offset)` pair, computed from the ordered
 * chapter durations (`game_sources.duration_s`). It is the shared contract with
 * the pipeline worker; the same rules must hold on both sides, so the boundary
 * semantics below are deliberate and load-bearing.
 *
 * Units: every value is seconds. Durations may be fractional.
 */

/**
 * A point inside a specific chapter file: `localOffsetS` seconds into the file
 * at `sourceIndex` (0-based, following `game_sources.order_index`).
 */
export interface SourcePoint {
  readonly sourceIndex: number;
  readonly localOffsetS: number;
}

/**
 * Validates that `durationsS` is a usable ordered chapter layout: a non-empty
 * list of finite, strictly-positive durations. A zero or negative duration is a
 * corrupt `game_sources` row, not a boundary case, so we fail loudly.
 */
function assertValidDurations(
  durationsS: readonly number[],
): asserts durationsS is readonly number[] {
  if (durationsS.length === 0) {
    throw new RangeError("game-time map needs at least one chapter duration");
  }
  for (let i = 0; i < durationsS.length; i += 1) {
    const duration = durationsS[i];
    if (!Number.isFinite(duration) || duration <= 0) {
      throw new RangeError(
        `chapter ${i} has a non-positive or non-finite duration (${duration}s)`,
      );
    }
  }
}

/**
 * Total length of the game in seconds: the sum of every chapter's duration.
 * Global offsets are valid in the closed interval `[0, totalDurationS]`.
 */
export function totalDurationS(durationsS: readonly number[]): number {
  assertValidDurations(durationsS);
  let total = 0;
  for (const duration of durationsS) {
    total += duration;
  }
  return total;
}

/**
 * Maps a global game-time offset to the chapter file that plays at that moment
 * and the local offset within it.
 *
 * Boundary rule (must match the pipeline worker): chapter intervals are
 * half-open `[start, start + duration)`, so an offset that lands exactly on an
 * interior boundary belongs to the *start* of the next chapter (local offset 0) -
 * that is where playback continues. The single exception is the game end: an
 * offset exactly equal to `totalDurationS` maps to the *end* of the last chapter
 * (`sourceIndex = N - 1`, `localOffsetS = last duration`), which is the natural
 * inverse for an exclusive clip end that reaches the final frame.
 *
 * @throws RangeError if `gameTimeS` is not finite or falls outside `[0, total]`.
 */
export function toSourcePoint(
  durationsS: readonly number[],
  gameTimeS: number,
): SourcePoint {
  assertValidDurations(durationsS);
  if (!Number.isFinite(gameTimeS) || gameTimeS < 0) {
    throw new RangeError(
      `game-time offset ${gameTimeS}s is out of range [0, total]`,
    );
  }

  let chapterStartS = 0;
  for (let index = 0; index < durationsS.length; index += 1) {
    const duration = durationsS[index];
    if (gameTimeS < chapterStartS + duration) {
      return { sourceIndex: index, localOffsetS: gameTimeS - chapterStartS };
    }
    chapterStartS += duration;
  }

  // Fell through the half-open scan: only the exact game end is still valid.
  if (gameTimeS === chapterStartS) {
    const lastIndex = durationsS.length - 1;
    return { sourceIndex: lastIndex, localOffsetS: durationsS[lastIndex] };
  }
  throw new RangeError(
    `game-time offset ${gameTimeS}s exceeds total ${chapterStartS}s`,
  );
}

/**
 * Maps a `(source file, local offset)` pair back to a global game-time offset:
 * the exact inverse of {@link toSourcePoint}. The local offset may equal the
 * chapter's duration (its exclusive end), which is how a boundary-crossing or
 * game-ending clip point is expressed.
 *
 * @throws RangeError if `sourceIndex` is not a valid chapter, or `localOffsetS`
 *   is not finite or falls outside `[0, chapter duration]`.
 */
export function toGameTime(
  durationsS: readonly number[],
  point: SourcePoint,
): number {
  assertValidDurations(durationsS);
  const { sourceIndex, localOffsetS } = point;
  if (!Number.isInteger(sourceIndex) || sourceIndex < 0) {
    throw new RangeError(`source index ${sourceIndex} is not a chapter`);
  }
  if (sourceIndex >= durationsS.length) {
    throw new RangeError(
      `source index ${sourceIndex} is out of range (${durationsS.length} chapters)`,
    );
  }
  const duration = durationsS[sourceIndex];
  if (!Number.isFinite(localOffsetS) || localOffsetS < 0) {
    throw new RangeError(
      `local offset ${localOffsetS}s is out of range [0, ${duration}]`,
    );
  }
  if (localOffsetS > duration) {
    throw new RangeError(
      `local offset ${localOffsetS}s exceeds chapter ${sourceIndex} duration ${duration}s`,
    );
  }

  let chapterStartS = 0;
  for (let index = 0; index < sourceIndex; index += 1) {
    chapterStartS += durationsS[index];
  }
  return chapterStartS + localOffsetS;
}

File: mac/HockeyKit/Sources/HockeyCore/GameTimeMap.swift
/// Global game time (ADR 0002): one coordinate in seconds since the start of the
/// game, mapped to a `(chapter, local offset)` pair over the ordered chapter
/// durations (`game_sources.duration_s`).
///
/// Port of `src/lib/time-mapping/game-time-map.ts`, pinned by
/// `contracts/vectors/time-mapping.json`. The boundary rule is load-bearing and
/// must match the web app: chapters are half-open `[start, start + duration)`, so
/// a time on an interior seam belongs to the next chapter, and only the exact
/// game end maps to the end of the last chapter.
///
/// Chapter starts are always sums of the durations in chapter order, the same
/// additions the TypeScript makes, so both apps agree on every seam bit for bit.

/// A point inside one chapter file: `localOffsetS` seconds into the chapter at
/// `sourceIndex` (0-based, in play order).
public struct SourcePoint: Equatable, Sendable {
    public let sourceIndex: Int
    public let localOffsetS: Double

    public init(sourceIndex: Int, localOffsetS: Double) {
        self.sourceIndex = sourceIndex
        self.localOffsetS = localOffsetS
    }
}

/// Why a game-time rule rejected its input. The rules reject rather than clamp,
/// because a bad layout or an out-of-range time is a bug upstream.
public enum GameTimeError: Error, Equatable, Sendable {
    /// The layout has no chapter.
    case noChapters
    /// A chapter's duration is zero, negative or not finite: a corrupt row.
    case invalidDuration(chapter: Int)
    /// A game time is not finite or lies outside `[0, total]`.
    case timeOutOfRange
    /// A chapter index names no chapter.
    case chapterOutOfRange
    /// A local offset is not finite or lies outside `[0, chapter duration]`.
    case offsetOutOfRange
    /// A window is not finite, not strictly ordered, or outside the game.
    case invalidWindow
}

/// Checks that `durationsS` is a usable layout: at least one chapter, each
/// finite and strictly positive.
func validateDurations(_ durationsS: [Double]) throws(GameTimeError) {
    if durationsS.isEmpty { throw .noChapters }
    for (index, duration) in durationsS.enumerated()
    where !duration.isFinite || duration <= 0 {
        throw .invalidDuration(chapter: index)
    }
}

/// The length of the game: the sum of every chapter's duration. Game times are
/// valid in the closed interval `[0, totalDurationS]`.
public func totalDurationS(_ durationsS: [Double]) throws(GameTimeError) -> Double {
    try validateDurations(durationsS)
    var total = 0.0
    for duration in durationsS {
        total += duration
    }
    return total
}

/// The chapter that plays at `gameTimeS` and the offset within it.
public func toSourcePoint(
    _ durationsS: [Double],
    gameTimeS: Double
) throws(GameTimeError) -> SourcePoint {
    try validateDurations(durationsS)
    guard gameTimeS.isFinite, gameTimeS >= 0 else { throw .timeOutOfRange }

    var chapterStartS = 0.0
    for (index, duration) in durationsS.enumerated() {
        if gameTimeS < chapterStartS + duration {
            return SourcePoint(sourceIndex: index, localOffsetS: gameTimeS - chapterStartS)
        }
        chapterStartS += duration
    }

    // Fell through the half-open scan: only the exact game end is still valid.
    if gameTimeS == chapterStartS {
        let lastIndex = durationsS.count - 1
        return SourcePoint(sourceIndex: lastIndex, localOffsetS: durationsS[lastIndex])
    }
    throw .timeOutOfRange
}

/// The game time of a point in a chapter: the exact inverse of `toSourcePoint`.
/// The local offset may equal the chapter's duration, its exclusive end.
public func toGameTime(
    _ durationsS: [Double],
    point: SourcePoint
) throws(GameTimeError) -> Double {
    try validateDurations(durationsS)
    guard durationsS.indices.contains(point.sourceIndex) else { throw .chapterOutOfRange }
    let duration = durationsS[point.sourceIndex]
    guard point.localOffsetS.isFinite, point.localOffsetS >= 0,
          point.localOffsetS <= duration
    else { throw .offsetOutOfRange }

    var chapterStartS = 0.0
    for index in 0..<point.sourceIndex {
        chapterStartS += durationsS[index]
    }
    return chapterStartS + point.localOffsetS
}

/// The game time each chapter starts at: chapter `i` starts at the sum of the
/// durations before it, added in chapter order.
public func chapterStartsS(_ durationsS: [Double]) throws(GameTimeError) -> [Double] {
    try validateDurations(durationsS)
    var starts: [Double] = []
    starts.reserveCapacity(durationsS.count)
    var chapterStartS = 0.0
    for duration in durationsS {
        starts.append(chapterStartS)
        chapterStartS += duration
    }
    return starts
}

/// A game time clamped into `[0, total]`, for seeks that overshoot either end
/// (a skip past the end, a step before the start). Not finite reads as 0, like
/// `clampGameTimeS` in `src/features/player/playback-plan.ts`.
public func clampGameTimeS(_ durationsS: [Double], _ gameTimeS: Double) throws(GameTimeError)
    -> Double
{
    let total = try totalDurationS(durationsS)
    guard gameTimeS.isFinite else { return 0 }
    return min(max(gameTimeS, 0), total)
}
